Correct broken RH control in UnitarySystem

rraustad opened the pull request:
Pull request overview
---------------------
 - Fixes #7782
 - Refactoring inadvertently declared global MoistureLoad variable as a local
 - Additional testing revealed other necessary changes were required

A new warning was added to inform users when inconsistent inputs are used for moisture control. When the AirloopHVAC:UnitarySystem uses Dehumidification Control Type = CoolReheat or Multimode, the Latent Load Control input should not be blank or specify SensibleOnlyLoadControl. Example files were updated according to these rules.

Load based controls in UnitarySystem were also improved when the no load air flow rate was greater than the low speed flow rate for multi-speed cooling coils. Controls typically failed for constant fan systems when the zone thermostat was in the deadband and operation of the fan would add sufficient load to push the zone temperature out of the deadband.

Example files were also updated to replace a blank latent load control field with a valid value when the dehumidification control type was equal to CoolReheat or Multimode and Control Type = Load.

    AirLoopHVAC:UnitarySystem,
      Load,                    !- Control Type
      CoolReheat,              !- Dehumidification Control Type
      ,                        !- Latent Load Control

Using:

    LatentWithSensibleLoadControl, !- Latent Load Control

Updated files are:

 - UnitarySystem_FurnaceWithDXSystemRHcontrol.idf
 - UnitarySystem_TwoStageDX
 - UnitarySystem_TwoStageDXWithHumidityControl

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [x] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [x] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [x] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [x] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] ~If any defect files are updated to a more recent version, upload new versions here or on
DevSupport~
 - [ ] ~If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label~
 - [ ] ~If structural output changes, add to output rules file and add OutputChange label~
 - [ ] ~If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies~

### Reviewer
This will not be exhaustively relevant to every PR.
 - [x] Perform a Code Review on GitHub
 - [x] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] ~If feature, test running new feature, try creative ways to break it~
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [x] Run Unit Test(s) locally
 - [ ] ~Check any new function arguments for performance impacts~
 - [ ] ~Verify IDF naming conventions and styles, memos and notes and defaults~
 - [ ] ~If new idf included, locally check the err file and other outputs~

rraustad committed:
Correct broken RH control in UnitarySystem

rraustad commented on src/EnergyPlus/UnitarySystem.cc:
```diff
@@ -7308,7 +7308,7 @@ namespace UnitarySystems {
 
         {
-            Real64 MoistureLoad = 0.0;
+            MoistureLoad = 0.0;
```
Big refactor oops

rraustad commented on tst/EnergyPlus/unit/UnitarySystem.unit.cc:
```diff
@@ -4124,6 +4125,14 @@ TEST_F(EnergyPlusFixture, UnitarySystemModel_GetInput)
+    thisSys->m_RunOnLatentLoad = true;
+    thisSys->simulate(thisSys->Name, FirstHVACIteration, AirLoopNum, CompIndex, HeatActive, CoolActive,
+        ZoneOAUnitNum, OAUCoilOutTemp, ZoneEquipment, sensOut, latOut);
+    EXPECT_LT(thisSys->m_MoistureLoadPredicted, 0.0); // dehumidification control type = CoolReheat so MoistureLoad < 0
```
Tests that MoistureLoad, saved in m_MoistureLoadPredicted is < 0 since RemainingOutputReqToDehumidSP = -0.00001;

rraustad commented:
Win64 results are bogus because of delay in results with multiple commits prior to CI execution

mjwitte committed:
Merge remote-tracking branch 'remotes/origin/develop' into 7782-AirloopHVACUnitarySystem-RH-control-broken-with-V92

mjwitte reviewed (commented):
@rraustad Codes changes are simple enough. Verified that the revised unit test fails with develop, passes here. Would like to see this in a full simulation. What idf would show the control change?

I've also merged in develop and will let CI do another pass.

rraustad commented:
Well, it was a Trane file that found this, and those don't run in E+. Let me see if I can find it and modify for E+ (yes I found it). Alternately, a UintarySystem file with CoolReheat and a supp heating coil and LatentOrSensibleLoadControl (or give me some time).

rraustad commented:
I ran the UnitarySystem CoolReheat file in develop and this branch.

![image](https://user-images.githubusercontent.com/5778156/75799142-a1fc1400-5d45-11ea-8734-c4ab51c528ea.png)

rraustad commented:
Defect file [here](https://github.com/NREL/EnergyPlusDevSupport/tree/master/DefectFiles/7000s/7782)

mjwitte commented:
Soo, why isn't UnitarySystem CoolReheat showing CI diffs? Is it because no diffs arise on the design days?

rraustad commented:
Because there is no example file named UnitarySystem_CoolReheat, that's just the name I gave the defect file. And there are no RH control example files for UnitarySystem (or unit tests) so no diff's

rraustad commented:
You gonna force me to create an example file before this gets checked in? I would.

mjwitte commented:
Seems that one or more of the new UnitarySystem example files from the coil refactor work should be showing diffs here.

mjwitte commented:
Like UnitarySystem_FurnaceWithDXSystemRHcontrol

rraustad commented:
That's a very good point. Latent load control is blank in that file and default is SensibleOnlyLoadControl, that should be corrected here.

mjwitte commented:
Would be good for unitarysystem to warn if CoolReheat or Multimode is active, but this field is SensibleOnlyLoadControl.

rraustad commented:
New warning from exisiting example file, comments? Make severity a warning? Change text?

    ** Severe  ** AirLoopHVAC:UnitarySystem = GASHEAT DXAC FURNACE 1
    **   ~~~   ** Latent Load Control = SensibleOnlyLoadControl
    **   ~~~   ** Dehumidification Control Type = COOLREHEAT
    **   ~~~   ** Humidity/Moisture will not be controlled with these settings

UnitarySystem_FurnaceWithRHControl without (existing) and with latent load control using LatentWithSensibleLoadControl

![image](https://user-images.githubusercontent.com/5778156/75807151-e5a94a80-5d52-11ea-87c5-ced5e058ce7e.png)

rraustad commented:
This is ready for check in but I only want one commit. Unit tests still run to completion. I revised the warning as:

    ** Warning ** AirLoopHVAC:UnitarySystem = DXAC HEAT PUMP 1
    **   ~~~   ** Inconsistent moisture control inputs.
    **   ~~~   ** Dehumidification Control Type = COOLREHEAT
    **   ~~~   ** Latent Load Control = SensibleOnlyLoadControl
    **   ~~~   ** Humidity/Moisture will not be controlled with these settings.

mjwitte commented:
I would vote for warning - someone might want to quickly compare the three options without having to change other stuff.

rraustad committed:
Added warning and correct example file

rraustad committed:
Merge branch '7782-AirloopHVACUnitarySystem-RH-control-broken-with-V92' of https://github.com/NREL/EnergyPlus into 7782-AirloopHVACUnitarySystem-RH-control-broken-with-V92

rraustad commented:
@mjwitte the new warning points to 5 other ill-formed example files with the same problem (sensible only control using CoolReheat). Fix those also?

rraustad commented:
I thought that there was no difference in results at first and then I noticed that the HX is on after correcting the example file.

![image](https://user-images.githubusercontent.com/5778156/75843408-12844e80-5da1-11ea-9bf0-c6f5be9e84ae.png)

rraustad commented:
For this example file I don't see any change in zone conditions when latent load control is corrected. This file is also set point based control.

![image](https://user-images.githubusercontent.com/5778156/75845028-da333f00-5da5-11ea-989f-caa8221e8268.png)

rraustad commented:
It may be that CoolReheat is ignored for set point control (which means warning may not be needed for SP control) and the coils are controlled solely by SPs. But there is a max humrat SP at the cooling coil outlet = 0.006 and it looks like that isn't being honored or controlled well or something. Air flow autosizes out to 5.3 m3/s for this 1-speed VS coil so there is plenty of room for increased capacity/latent.

![image](https://user-images.githubusercontent.com/5778156/75845793-3a2ae500-5da8-11ea-9afe-c47f7a8f7506.png)

rraustad commented:
Load controlled with DX coil Alt modes using Multimode control with blank Latent Load Control field (and changing RH to 50%) and then using LatentWithSensibleLoadControl, no difference. This is wrong.

![image](https://user-images.githubusercontent.com/5778156/75881742-a845cb00-5ded-11ea-8c12-4c12742b84e1.png)

mjwitte commented:
@rraustad I would say, yes, we need to fix all of those example files and for ones that have an equivalent in the coil refactor branch, we need to fix those over there. I'll wait for you to sort out the ones here that don't control well. Or let me know if you want to fork those off to another PR.

mjwitte commented:
@rraustad So, where do we stand on this PR?

rraustad commented:
I'll check where I was before getting pulled away. Should be close to ready.

rraustad committed:
Correct example files and changes to control code for VS coils.

rraustad commented:
OK, I fixed the example files, modified the warning message and added additional controls for VS coils. This is creeping out of control so other things found should go in a new issue. The zone RH profile still looks like it did 8 days ago [here](https://github.com/NREL/EnergyPlus/pull/7783#issuecomment-594053734). Boy does time fly.

rraustad committed:
Correct warning

rraustad committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into 7782-AirloopHVACUnitarySystem-RH-control-broken-with-V92

rraustad commented:
Looks like UnitarySystem_TwoStageDXWithHumidityControl has some control issues once inputs were corrected. Follow up issue warranted?

rraustad commented:
I ran UnitarySystem_TwoStageDXWithHumidityControl, with Chicago Ohare (testfiles/CMakeLists.txt), Chicago IL (idf location), pulling latest on this branch, and checking for correct weather file and example file, I don't see those new control warnings ?? @mjwitte can you duplicate?

Chicago-Ohare:

    Program Version,EnergyPlus, Version 9.3.0-a06c012ddf, YMD=2020.03.12 08:53,
    ************* Beginning Zone Sizing Calculations
    ************* Beginning System Sizing Calculations
    ************* Testing Individual Branch Integrity
    ************* All Branches passed integrity testing
    ************* Testing Individual Supply Air Path Integrity
    ************* All Supply Air Paths passed integrity testing
    ************* Testing Individual Return Air Path Integrity
    ************* All Return Air Paths passed integrity testing
    ************* No node connection errors were found.
    ************* Beginning Simulation
    ************* Simulation Error Summary *************
    ************* EnergyPlus Warmup Error Summary. During Warmup: 0 Warning; 0 Severe Errors.
    ************* EnergyPlus Sizing Error Summary. During Sizing: 0 Warning; 0 Severe Errors.
    ************* EnergyPlus Completed Successfully-- 0 Warning; 0 Severe Errors; Elapsed Time=00hr 00min  1.58sec

rraustad commented:
I turned on Run Simulation for Sizing Periods and now I see the warnings.

mjwitte commented:
@rraustad Regarding UnitarySystem_TwoStageDXWithHumidityControl, it's your call whether to attempt a fix here or post a new issue.

rraustad commented:
Trying to see if I can find a solution, if not I'll post a new issue. I'll know which by later today.

rraustad commented:
I got rid of the control errors but am still absorbing what controls are correct. For this 2-spd coil the idle flow is 0.22 and speed 1 flow is 0.11 and speed 2 flow is 0.22. So the idle flow is full flow which I have questioned in the past. Shouldn't the idle flow be <= low speed flow?

When the control error occurred, the flow rate was set at idle flow and output was 470 W which pushed the zone T out of the deadband above the cooling SP temp. So the Init code said the load was = load to cooling SP, which was 298 W.

Then in the control routine it runs the system at PLR = 0 and idle flow which gives 470 W. Then it activates speed 1 flow and no compressor outlet = 235 W. Had I gone to speed 2 I could have hit 298 W with coil active, but what should happen here? Should I exceed the requested load using speed 1 flow or go to speed 2 and use the compressor? 

This figure is with and without the 2 lines of code (7838 and 7844) that corrected the control error. Results are the same except that the control warnings are gone with these changes.

![image](https://user-images.githubusercontent.com/5778156/76549134-3a8d4500-6466-11ea-83ba-7d2b9bd1455a.png)

rraustad committed:
Correct controls to eliminate control warnings

mjwitte commented:
@rraustad At this point, I would change the idf to make the idf flow rate <= speed 1 flow rate.

rraustad commented:
I'm having a problem back tracking this variable. Looks to be a code change and not an idf change. I think idle flow was added to the parent for the VS coil (not sure why since there was already a no load flow rate variable). If I look back at the Furnace code, IdleVolumeAirRate is not an input, it's a resultant of the variable speed flows. And it is the lowest speed flow in Furnace. So I think, in UnitarySystem, this variable was treated like MaxNoCoolHeatAirVolFlow, which was autosized to the single speed coil flow, and when VS coils were introduced, autosized to the max speed flow. So yes, this will need to change for all VS/MS coil types if the autosized value is changed to the minimum speed flow. I do agree with this.

    if (Furnace(FurnaceNum).NumOfSpeedHeating > 0) {
        Furnace(FurnaceNum).IdleMassFlowRate = min(Furnace(FurnaceNum).HeatMassFlowRate(1), Furnace(FurnaceNum).CoolMassFlowRate(1));
        Furnace(FurnaceNum).IdleSpeedRatio = min(Furnace(FurnaceNum).MSHeatingSpeedRatio(1), Furnace(FurnaceNum).MSCoolingSpeedRatio(1));
        Furnace(FurnaceNum).IdleVolumeAirRate = min(Furnace(FurnaceNum).HeatVolumeFlowRate(1), Furnace(FurnaceNum).CoolVolumeFlowRate(1));
    } else {
        Furnace(FurnaceNum).IdleMassFlowRate = Furnace(FurnaceNum).CoolMassFlowRate(1);
        Furnace(FurnaceNum).IdleSpeedRatio = Furnace(FurnaceNum).MSCoolingSpeedRatio(1);
        Furnace(FurnaceNum).IdleVolumeAirRate = Furnace(FurnaceNum).CoolVolumeFlowRate(1);
    }

rraustad commented:
Here it is in UniarySystem

    if (MSHPIndex > 0) {
        this->m_IdleVolumeAirRate = this->m_MaxCoolAirVolFlow * designSpecMSHP[MSHPIndex].noLoadAirFlowRateRatio;
    } else if (this->m_CoolVolumeFlowRate.size() == 0) {
        this->m_IdleVolumeAirRate = this->m_MaxNoCoolHeatAirVolFlow;
    }

rraustad commented:
Max iterations for UnitarySystem_TwoStageDXWithHumidityControl.idf are gone when OA flow is set to 0.

    Controller:OutdoorAir,
      0,                !- Minimum Outdoor Air Flow Rate {m3/s}
      0,                !- Maximum Outdoor Air Flow Rate {m3/s}


    Program Version,EnergyPlus, Version 9.3.0-a06c012ddf, YMD=2020.03.12 20:03,
    ************* Beginning Zone Sizing Calculations
    ************* Beginning System Sizing Calculations
    ************* Testing Individual Branch Integrity
    ************* All Branches passed integrity testing
    ************* Testing Individual Supply Air Path Integrity
    ************* All Supply Air Paths passed integrity testing
    ************* Testing Individual Return Air Path Integrity
    ************* All Return Air Paths passed integrity testing
    ************* No node connection errors were found.
    ************* Beginning Simulation
    ************* Simulation Error Summary *************
    ************* EnergyPlus Warmup Error Summary. During Warmup: 0 Warning; 0 Severe Errors.
    ************* EnergyPlus Sizing Error Summary. During Sizing: 0 Warning; 0 Severe Errors.
    ************* EnergyPlus Completed Successfully-- 0 Warning; 0 Severe Errors; Elapsed Time=00hr 00min  2.34sec

mjwitte commented:
@rraustad Any progress here to wrap this up?

rraustad commented:
I got as close as possible given existing code. I think modifying existing code to change/correct something should be done in a new issue. The underlying issue was resolved, example files were updated, a new warning added and VS coil controls were corrected where I saw a problem. There is issue creep here and these changes should be sufficient for this issue/PR.

rraustad commented:
Added issue #7862.

Myoldmopar reviewed (approved):
No problems seen, and glad to see an issue opened to continue the effort later.

Myoldmopar commented:
Merge at will.

mjwitte commented:
@Myoldmopar I haven't done any local testing of the final result here.
Also, any idf changes here may need to be replicated in dx coil/baseline branch?

Myoldmopar commented:
I'm not sure if the IDF changes will need to be propagated anywhere else.  I mean if they are already in the UnitarySystem files here, I don't know what else we'd need to do besides just pull in the develop branch.  

As for final testing, that would be great, but I don't intend on spending further time on it right now.

Myoldmopar commented:
@mjwitte I will make the call to merge, but I'm hesitant because it seemed like maybe you wanted a little more time with this.  I'm fine either way.

mjwitte committed:
Merge remote-tracking branch 'remotes/origin/develop' into 7782-AirloopHVACUnitarySystem-RH-control-broken-with-V92

mjwitte reviewed (approved):
Tested the defect file, 7782-UnitarySystem_CoolReheat, max/min/average temperatures are similar (average is a bit colder as expected with more dehumidification), max/min/average RH is much improved.

Changed idf and saw the new "Inconsistent moisture control inputs" warning.

Code changes look reasonable.

Pulled in develop and ran all UnitarySystem unit tests.

Will push up and let CI run since it's been a few days for this branch.

mjwitte commented:
Oh, but wait, expandobjects is not writing the latent load control field, will check in a fix shortly.

mjwitte committed:
Fix HVACTemplate UnitarySystem Latent Load Control field

mjwitte commented:
There's one unit test failure that appears to be unrelated to this PR. Merging this and we'll have to clean up unit tests with a hotfix.
